keps/sig-node/5307-container-restart-policy/README.md

```diff
@@ -0,0 +1,1052 @@
+aligns with the goal of expediting in-place container restart. First, the
+in-place restart avoids the expensive Pod re-scheduling to a different node.
+Second, if the container keeps restarting due to an exit code specified in
+the restart rules and stuck in a CrashLoop, it is probably not a retry-able
+error, the exponential backoff can avoid overwhelming the node with frequent
+restarts.
+
+## Design Details
+
+The proposal is to extend the Pod's [Container API type](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/types.go#L2528) 
+with a new field `ContainerRestartrRules`. A list of rules will be evaluate
+in order to determine the action of container exit based on the conditions.
+The initial proposal supports only one action, "Restart", to restart the container.
+The initial proposal supports only exit code as requirement for the rules.
+
+The proposed API is as following:
+
+```go
+type Container struct {
+  // Omitting irrelevant fields...
+
+  // Represents a list of rules to be checked to determine if the
+  // container should be restarted on exit. The rules are evaluated in
+  // order. Once a rule matches a container exit condition, the remaining
+  // rules are ignored. If no rule matches the container exit condition,
+  // the Pod-level restart policy determines the whether the container
+  // is restarted or not.
+  RestartRules []ContainerRestartRule
+}
+
+// ContainerRestartRule describes how a container exit is handled.
+type ContainerRestartRule struct {
+  // Specifies the action taken on a container exit if the requirements
+  // are satisfied. The only possible value is "Restart" to restart the
+  // container.
+  Action ContainerRestartRuleAction
+  
+  // Represents the requirement on the container exit codes.
+  OnExitCodes *ContainerRestartRuleOnExitCodesRequirement
+}
+
+type ContainerRestartRuleAction string
+
+const (
+  ContainerRestartRuleActionRestart ContainerRestartRuleAction = "Restart"
+)
+
+// ContainerRestartRuleOnExitCodesRequirement describes the requirement
+// for handling an exited container based on its exit codes.
+type ContainerRestartRuleOnExitCodesRequirement struct {
+  // Represents the relationship between the container exit code(s) and the
+	// specified values. Possible values are:
+	//
+	// - In: the requirement is satisfied if the container exit code is in the 
+  //   set of specified values.
+	// - NotIn: the requirement is satisfied if the container exit code is 
+  //   not in the set of specified values.
+  Operator ContainerRestartRuleOnExitCodesOperator
+
+  // Specifies the set of values to check for container exit codes.
+  // At most 255 elements are allowed.
+  Values []int32
+}
+
+type ContainerRestartRuleOnExitCodesOperator string
+
+const (
+  ContainerRestartRuleOnExitCodesOpIn ContainerRestartRuleOnExitCodesOperator = "In"
+  ContainerRestartRuleOnExitCodesOpNotIn ContainerRestartRuleOnExitCodesOperator = "NotIn"
```

Comment:
personally I think listing the codes would cover the cases in a clearer way but I am open to this.

Reply:
hm actually I see what you are doing here and I take it back, the in/not in language is fine I think.